### ISSUE TITLE
Support for more variable types in yaml default files

### ISSUE THREAD

**Issue by mjfs** [jgm/pandoc#7253]
With default files (via `--defaults` switch) only a limited set of environment variables is available for use in path related fields. A really common use case where one would convert `markdown` file to `docx` file with only extension being different in the filename (apart from path) is therefore not supported.

Supporting more variable types, similarly to how this is implemented for `Lua` filters via `CommonState`, would improve user experience since even fewer switches and logic for achieving desired behavior would be needed. By supporting access to the name of the first input file without extension, passed via command line, the following would be possible:

`pandoc --defaults=my_defaults.yaml my_document.md`

After executing only the command above file `my_document.docx` could be created in `TEMP` subfolder of the folder where `my_document.md` is located.

**Comment by jgm** [comment]
So currently we have
```
${HOME}
${USERDATA}
${.}
```
in field that expect file paths.  You would like to add
```
${BASENAME}
```
or something like that, based on the filename of the first input file.

**Comment by mjfs** [comment]
Yes, `BASENAME` (e.g. of the first input file if there are many) is exactly the scenario in the example above. Other variables could also be useful (e.g. author, date/timestamp), but the name without an extension probably already covers a lot and would suffice in my current use case.

**Comment by wlupton** [comment]
I hope it's OK if I hijack this issue for further discussion of the same topic. Two things:

1. The manual says "*the following syntax may be used to interpolate environment variables*" which, until I read this issue just now, I assumed referred to _all_ environment variables (with the caveat that for `USERDATA` the actual user data directory trumps the environment variable). I now realise that it's not referring to environment variables in general. Perhaps this could be clarified?

2. The term "*only works in fields that expect file paths*" is rather vague, and it might be good to have a list of these fields? Note (BTW) that #8024 suggests that `defaults` isn't included in this list (perhaps unintentionally). But I wanted to substitute values in some PDF engine options and there's no way that pandoc can know which of these "expect file paths", so I'm wondering why the translation can't simply be applied to the entire file (as well as perhaps supporting all environment variables). Or is there a security concern?

**Comment by jgm** [comment]
No, if I recall correctly, all environment variables should work.

**Comment by wlupton** [comment]
So is https://github.com/jgm/pandoc/issues/7253#issuecomment-829368103 outdated?

**Comment by wlupton** [comment]
Thanks. I've checked, and of course you're correct! (I might also have been confused by the reference at the top to "a limited set of environment variables"). BTW, why single out `${HOME}`?

I created a defaults file to test most of the file-like settings and noted the following:

* `defaults` indeed doesn't appear to support the `${}` syntax, as noted in #8024 (is this a bug? I'll update that issue)
* It's not a good idea to set `data-dir` to `${USERDATA}`! (why would you... but should it result in an infinite loop?)
* As expected, `pdf-engine-opts` doesn't support the `${}` syntax (I'll create a new issue for this)

On the original topic of this issue, I'm not sure that I agree with supporting `${BASENAME}`. It seems like a slippery slope, and you can surely do something like this (depending on your shell):

```
BASENAME=$(basename $FILE) pandoc $FILE
```

However, I could imagine bash-like substitutions being supported in defaults files, but they are complicated and I always have to look them up! For example ([see here](https://tldp.org/LDP/abs/html/parameter-substitution.html#PSOREX2)), this does something like basename (it removes the longest match of the `*/` pattern from the beginning of the value):

```
echo ${FILE##*/}
```
